### PR TITLE
Fixed #25693 -- Prevented prefetched data and field collision

### PR DESCRIPTION
**Pull request by Ian-Foote**
Thanks to Jamie Matthews for finding and explaining the bug.

### CODE REVIEW

File: tests/prefetch_related/tests.py [hunk 1]
```diff
@@ -906,6 +907,24 @@ def test_m2m(self):
             )
 
 
+class ShadowTest(TestCase):
+    @ignore_warnings(category=RemovedInDjango20Warning)
```
Comment (timgraham) [inline tests/prefetch_related/tests.py]: which deprecation warning is being ignored?

File: tests/prefetch_related/tests.py [hunk 2]
```diff
@@ -906,6 +907,24 @@ def test_m2m(self):
...
+    def test_m2m_shadow(self):
+        book1 = Book.objects.create(title='Poems')
+        book2 = Book.objects.create(title='Code')
+        author = Author.objects.create(name='Jane', first_book=book1)
+        author.books.add(book1)
+        author.books.add(book2)
+
+        poems = Book.objects.filter(title='Poems')
+        with self.assertRaises(ValueError):
```
Comment (timgraham) [inline tests/prefetch_related/tests.py]: please use assertRaisesMessage to check the message too.

File: tests/prefetch_related/tests.py [hunk 3]
```diff
@@ -906,6 +907,24 @@ def test_m2m(self):
...
 
 
+class ShadowTest(TestCase):
+    @ignore_warnings(category=RemovedInDjango20Warning)
+    def test_m2m_shadow(self):
+        book1 = Book.objects.create(title='Poems')
+        book2 = Book.objects.create(title='Code')
+        author = Author.objects.create(name='Jane', first_book=book1)
+        author.books.add(book1)
+        author.books.add(book2)
```
Comment (charettes) [inline tests/prefetch_related/tests.py]: author.books.set([book1, book2])

File: tests/prefetch_related/tests.py (hunk 1 above)
Comment (Ian-Foote) [inline tests/prefetch_related/tests.py]: https://github.com/django/django/blob/9c5e272860c076736237d4b280c7c922c46ba273/django/db/models/fields/related_descriptors.py#L482-L486

See also: https://code.djangoproject.com/ticket/25550

Comment (charettes) [inline tests/prefetch_related/tests.py]: Are you sure this code path is still triggered now that you raise a `ValueError`?

File: tests/prefetch_related/tests.py (hunk 3 above)
Comment (Ian-Foote) [inline tests/prefetch_related/tests.py]: I decided not to do this to make this easier to backport to 1.8.

Comment (charettes) [inline tests/prefetch_related/tests.py]: Good consideration!

File: tests/prefetch_related/tests.py (hunk 1 above)
Comment (Ian-Foote) [inline tests/prefetch_related/tests.py]: Good point.

**Comment by charettes** [comment]
This looks like it's only missing a release note for 1.7 and 1.8.

File: django/db/models/query.py [hunk 4]
```diff
@@ -1575,6 +1575,17 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
...
         to_attr, as_attr = lookup.get_current_to_attr(level)
+
+        # Check we are not shadowing a field on obj
+        if as_attr:
+            try:
+                field = obj._meta.get_field(to_attr)
+            except exceptions.FieldDoesNotExist:
+                pass
+            else:
+                msg = 'to_attr={} shadows the {} field.'.format(to_attr, field)
```
Comment (j4mie) [inline django/db/models/query.py]: I'm not sure I'm a big fan of this error message. "Shadowing" has a specific meaning that may not be clear to everyone. Perhaps something along the lines of "You cannot use the name ... as the to_attr argument, as the model already has a field with this name"

Comment (charettes) [inline django/db/models/query.py]: What about simply replacing `shadows` with `conflicts` instead?

**Comment by Ian-Foote** [comment]
I've added release notes and tweaked the exception message.

**Comment by timgraham** [comment]
merged in 4608573788c04fc047da42b4b7b48fdee8136ad3, thanks!

**Comment by timgraham** [comment]
The fix doesn't work on 1.7 due to API changes in `get_field()` I guess. Ian or Simon, do you guys care about fixing it there? Personally, I don't care since 1.7 will be EOL 3 weeks from today.

**Comment by charettes** [comment]
From a quick look it looks like we would also need to call `get_all_related_objects` on Django 1.7 to get reverse fields. I'll see if I can find time to get this sorted today.

**Comment by timgraham** [comment]
Feel free to use https://github.com/timgraham/django/tree/25693-django17 as a starting point.

**Comment by charettes** [comment]
Thanks Tim!

**Comment by charettes** [comment]
See #5624

**Pull request closed without merging**